### ISSUE TITLE
Fine tune yolov8 custom trained model

### ISSUE THREAD

**Issue by sujonahmed2500** [ultralytics/ultralytics#5135]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I trained yolov8 from almost 5,000 pictures by google colab.

My colab file is: [Yolov8.ipynb](https://drive.google.com/file/d/1TvBelQ6Ww2q5EFLOtaTGVN3BvL8Q16b4/view?usp=sharing)
My data.yaml file: [data.yaml](https://drive.google.com/file/d/15HvpAx0CShGOQf94viCE0s_wtLyQUrbm/view?usp=sharing)

After the train is done it gives me a model named best.py and other files.

Now I want to train again this 'best.py' model with another 500 pictures.  
I have 500 pictures and labeled 500 .txt file.

How I can train my already trained model with new images?

Please help me. it's argent.

**Comment by RizwanMunawar** [comment]
Hello @sujonahmed2500, if you've previously trained a model and now want to fine-tune it on another set of 500 images, you can utilize the `best.pt` file from your previous training as a starting point. You can achieve this by using the following command:

```cli
yolo task=detect mode=train model="path/to/previous/train/best.pt" data=data.yaml
```

Thanks

**Comment by sujonahmed2500** [comment]
@RizwanMunawar thanks for helping me,
are you are about this?
let me explain my issue, first I train my 5000 picture like this:

`yolo detect train model=yolov8s.pt data=ata.yaml epochs=25 imgsz=640 plots=True`

After that, I noticed that I missed one class to label, so I labeled some new images for the missing class.
then I trained again like what you said. but after I got a new model it detected only the last class I labeled.

**Comment by RizwanMunawar** [comment]
Thank @sujonahmed2500, for the comprehensive explanation. It's important to note that you should annotate both classes in a single training session; you cannot split the training into individual classes.

The process is as follows: If you've previously trained a model on 5000 images, encompassing all the classes you intend to detect, and later you wish to incorporate additional data, which also includes annotations for all the target classes, you can then utilize the previously trained model for this purpose.

**Comment by sujonahmed2500** [comment]
@RizwanMunawar Sorry, I think I didn't understand you, 
classes are the same in every train, look my first training data has 25 classes defined in data.yaml file. in the first training data there is not enough data for class id 24. That's why I labeled some new images for class id 24 (also if images has other classes I also label them too). but when I train with this 

My colab file is: [Yolov8.ipynb](https://drive.google.com/file/d/1TvBelQ6Ww2q5EFLOtaTGVN3BvL8Q16b4/view?usp=sharing)
My data.yaml file: [data.yaml](https://drive.google.com/file/d/15HvpAx0CShGOQf94viCE0s_wtLyQUrbm/view?usp=sharing)

it only detects the last classes. not previous classes I trained.

**Comment by glenn-jocher** [comment]
@sujonahmed2500 hello, I understand your situation better now. When you're retraining the model, your training data needs to include examples of all classes you want the model to detect, not just the new class you want to add information of. So even if you run the training utilizing the previous model as a starting point, you have to include training examples of all classes - including the ones it had previously been trained on. 

This means that you should augment your new training data set to include examples of all the classes and not just the new data from class 24. After you've assembled this dataset (both old and new images inclusive), you can commence the retraining process. By incorporating examples of all classes you intend to detect, your newly trained model should be capable of recognizing instances of all those classes.

It is crucial that your `data.yaml` file is also updated to reflect all labels in the new training dataset. The model will only learn to distinguish classes it has been trained to recognize. It seems like it's currently only learning to detect the new class due to the absence of the other classes in the new training dataset.

**Comment by sujonahmed2500** [comment]
@glenn-jocher Thank you so much for helping me. 
How can I update my `data.yaml` file for new training.
Can you please suggest me something please?

**Comment by glenn-jocher** [comment]
Hello @sujonahmed2500, I'm glad you found the previous information helpful. 

To update your `data.yaml` file for the new training, you should include the paths to all of your training and validation images. This means you need to include the paths to both the old images as well as the new ones you've just labeled.

Also, you should make sure your `data.yaml` file correctly specifies the number of classes and their names.

Finally, updating the `data.yaml` file isn't storage intensive -- you are only specifying the paths to the images, not storing the images themselves in the file. After doing this, you'll be able to train your model on both the old and the new images.

**Comment by glenn-jocher** [comment]
Hello @sujonahmed2500,

In the `data.yaml` file, `train`, `val`, and `test` should each point to a directory where your training, validation, and testing images (respectively) are stored. To include the new data along with the old, all images (new and old) must be located in the appropriate folders as specified in the `data.yaml` file. 

So, you would place all the training images (both old and new) in the directory specified as the `train` path in the `data.yaml` file. The same applies to the validation and testing images – place both old and new images in the appropriate directories. 

Remember, your new training data should include the complete set of classes as defined in the `names` field in your `data.yaml` file. This will ensure that the new model you train can detect all the desired classes.

Hope this helps! Let us know if you need more clarification.

**Comment by sujonahmed2500** [comment]
Hello @glenn-jocher,
I store all my training images (old & new) in `train/images` and labels in `train/labels` folder. The same applies to the validation and testing images. 
Then I tried to train with the previous model by this command:

`yolo detect train data=data/data.yaml model=best08092023.pt epochs=25 imgsz=640`

`best08092023.pt` is my previously trained model.

but when I run it, it takes the same time as the first training takes.
Here is the screenshot:
![image](https://github.com/ultralytics/ultralytics/assets/72850751/5ffd576f-27e3-4c51-9f12-054f89d0a2b6)

**Comment by glenn-jocher** [comment]
Hi @sujonahmed2500,

Your new training run taking the same amount of time as the first one is to be expected. Although you're using a pre-trained model (in this case, `best08092023.pt`), the model still has to go through all the specified epochs you set (25 in this case) to fine-tune itself on the new dataset (old and new images).

This time-consuming task involves the model going over all of your training images, updating its weights based on its predictions, and slowly learning the details of your new dataset. Even when starting with a pre-trained model, this process of going through epochs still takes time.

So don't worry, this is a completely normal scenario. Although using the pre-trained model may not shorten the training time, it should improve the performance as it already has learnt features from your previous dataset. 

Hope this clarifies your concerns! Let us know if you have more questions.

**Comment by sujonahmed2500** [comment]
Thanks, everyone for helping me.
you are awesome.

**Comment by glenn-jocher** [comment]
You're welcome @sujonahmed2500! We are glad to help. Don't hesitate to ask if you have any other questions or come across any issues. We're always here to support. Happy developing with YOLOv8!

**Comment by ohadOrbach** [comment]
Hey
I didn't see any comment about changing initial parameters like lr, batches and more.
Also freezing some layers maybe.

Please share what more parts are relevant for this fine tuning

**Comment by glenn-jocher** [comment]
@ohadOrbach hello! For fine-tuning YOLOv8, you can adjust initial parameters such as learning rate (`lr`), batch size (`batch-size`), and others directly in the training command or in a configuration file. Freezing layers is not a typical practice with YOLOv8 as it automatically handles the transfer learning aspects. Focus on adjusting the data augmentation, learning rate, and epochs to suit your new dataset for optimal results. Check out the documentation for more on hyperparameter tuning. Happy fine-tuning! 🚀

**Comment by Asshok1** [comment]
Hey @glenn-jocher my major task is to detect small objects with high confidence score and IoU initially i have imgsz of 1280*1280 
for training but yolov8 resizes this to 640*640, as the resizing happens I am losing some information so what I think is instead of directly feeding 1280*1280  to model, I divided the image into 640*640 multiples and labelled and I need to train with these images. Is it a good practice or not ? and as I divided the imgsz into 640*640 before giving and trained, is it capable to detect small objects when new data is given ?

**Comment by glenn-jocher** [comment]
Hey there! 🙌

Dividing your 1280x1280 images into multiple 640x640 segments can be quite effective for detecting small objects, as it preserves the original resolution of those areas. This method is widely used and known as image tiling. It's a good practice, especially for scenarios demanding high precision on smaller objects.

When you train YOLOv8 on these 640x640 segments, the model learns from the higher-resolution fragments, thus becoming more capable of detecting small objects. Just make sure your labels are correctly adjusted for the new segments to maintain accuracy.

For future inferences, if you're working with images larger than what you trained on (e.g., 1280x1280 or larger), consider applying the same tiling strategy before feeding them to your model. This should help in maintaining detection performance on small objects.

Happy training! 🚀

**Comment by Asshok1** [comment]
@glenn-jocher Thank you so much

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions or need further assistance, feel free to ask. Happy coding! 😊

**Comment by SKaraboM** [comment]
Is it possible to add a new class to yolov8n-post.pt (Pre-trained model), without removing/losing the class that it was initially trained on?

**Comment by glenn-jocher** [comment]
@SKaraboM absolutely! You can add new classes to your YOLOv8n model without losing the previously trained classes by combining the datasets (old and new classes) and continuing the training. This way, the model learns to detect new classes while retaining its ability for the old ones. Just make sure to update your `data.yaml` to include all classes and adjust your training annotations accordingly. Here's a snippet for updating the YAML:

```yaml
nc: number_of_classes  # Update this to include your new class
names: ['class1', 'class2', ..., 'new_class']  # Add your new class at the end
```

Then, continue training as usual. Happy detecting! 🚀

**Comment by SKaraboM** [comment]
And what of I don't have the dataset the dataset that it was initially trained on. Does that mean there is no way to do it without the initial dataset?

**Comment by glenn-jocher** [comment]
Hey there! If you don't have the original dataset, you can still train the model with your new class by starting with the pre-trained weights. Just ensure your new dataset is sufficiently diverse and representative of the classes you want to detect, including the new ones. Here’s how you can set up your training:

```python
from ultralytics import YOLO

# Load the pre-trained model
model = YOLO('yolov8n-post.pt')

# Train with your new dataset
results = model.train(data='new_data.yaml')
```

This approach leverages the learned features from the pre-trained model while adapting to the new classes. Happy training! 😊

**Comment by SudhanshuBlaze** [comment]
Hi

I've a similar question regarding fine tuning of the yolov9 model. I'm using the ultralytics package for using yolov9. The model is pretty good at detecting and tracking the vehicles, but sometimes misclassifies some vehicles, like it sometimes detects buses as trains, and bigger cars as trucks. So I want to fine-tune the model on my dataset (video) to “classify” the vehicles correctly. I don't want to fine tune for detection or tracking as it is already good in it, so I want to only fine tune it for classifying the vehicle's correctly. 

So in simple words I want to fine tune the model for the same classes like bicycle and pedestrians. I don't want to add new classes. So in such case do I need label all the classes again in my custom training dataset or it would work even if I label only the bicycle and pedestrian classes. 

Please guide me with the steps or resources.

**Comment by glenn-jocher** [comment]
Hi there!

For fine-tuning your YOLOv9 model specifically for better classification accuracy on existing classes (bicycles and pedestrians), you can indeed focus on just these classes in your new dataset. You don't need to relabel all classes if your primary goal is to improve classification for specific categories.

Here’s a brief guide:
1. **Prepare your dataset** with labeled videos/images focusing on 'bicycle' and 'pedestrian' classes.
2. **Update your data.yaml** to ensure it only includes the classes you're focusing on.
3. **Continue training** using the pre-trained YOLOv9 weights. Make sure to adjust the learning rate and training epochs according to your dataset size to avoid overfitting.

This targeted approach should help improve classification accuracy for the specified classes without altering the model's existing capabilities in detection and tracking.

Happy fine-tuning! 😊

**Comment by SudhanshuBlaze** [comment]
Thanks for the prompt response, I will try this and let you know the results. Hopefully, the model does not lose its existing capability of detecting other classes.

**Comment by glenn-jocher** [comment]
You're welcome! I'm glad to help. Yes, please do keep us updated on your progress. If you carefully manage the training process, your model should maintain its proficiency in detecting other classes while improving on the ones you're focusing on. Looking forward to hearing about your results! 😊

**Comment by SudhanshuBlaze** [comment]
Hi, I am trying to fine-tune the model, using the below piece of code, the images which I am using are of `1920x1080` dimension, but by default the yolo scales it 640x640 which leads to lose in image data as there are small objects which I am trying to detect so I would like to keep the original image resolution while fine-tuning the model. Please help me achieve it.

```
from ultralytics import YOLO

yolo = YOLO('yolov9e.pt')
yolo.train(data='datasets/finetune_dataset/vehicles.yml', epochs=50, imgsz= 640)
valid_results = yolo.val()
print(valid_results)
```

**Comment by glenn-jocher** [comment]
Hi there! 👋

Thank you for reaching out. To fine-tune your model while preserving the original image resolution, you can adjust the `imgsz` parameter to match your image dimensions. This will ensure that the model processes the images at `1920x1080` without resizing them to `640x640`.

Here's how you can modify your code:

```python
from ultralytics import YOLO

# Load the pre-trained model
yolo = YOLO('yolov9e.pt')

# Train the model with the original image size
yolo.train(data='datasets/finetune_dataset/vehicles.yml', epochs=50, imgsz=(1920, 1080))

# Validate the model
valid_results = yolo.val()
print(valid_results)
```

This change should help you retain the image details necessary for detecting small objects. If you encounter any issues or have further questions, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example would be very helpful for us to investigate further. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Happy fine-tuning! 🚀

**Comment by shivam55sit** [comment]
how can we use Early Stopping while training YOLOv8x-cls.pt model ? Also, the data augmentation is handle by YOLO by itself or we can set that according to our needs?

**Comment by glenn-jocher** [comment]
Hi there!

Great questions! 😊

### Early Stopping
To implement early stopping while training your `YOLOv8x-cls.pt` model, you can set the `patience` parameter in your training configuration. This parameter determines how many epochs to wait for an improvement in validation metrics before stopping the training. Here’s an example:

```python
from ultralytics import YOLO

# Load the pre-trained model
model = YOLO('yolov8x-cls.pt')

# Train the model with early stopping
model.train(data='path/to/data.yaml', epochs=100, patience=5)
```

In this example, training will stop if there is no improvement in validation metrics for 5 consecutive epochs.

### Data Augmentation
YOLOv8 handles data augmentation by default, but you can customize it according to your needs. You can set various augmentation parameters in your training configuration file or script. Here’s an example of how to customize data augmentation:

```yaml
# data.yaml
train: path/to/train/images
val: path/to/val/images

# Number of classes
nc: 10

# Class names
names: ['class1', 'class2', ..., 'class10']

# Augmentation settings
augment:
  hsv_h: 0.015  # Adjust hue
  hsv_s: 0.7    # Adjust saturation
  hsv_v: 0.4    # Adjust value
  degrees: 10.0 # Rotate
  translate: 0.1 # Translate
  scale: 0.5    # Scale
  shear: 2.0    # Shear
  flipud: 0.0   # Flip up-down
  fliplr: 0.5   # Flip left-right
```

You can also set these parameters directly in your training script:

```python
model.train(data='path/to/data.yaml', epochs=100, augment={'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 10.0, 'translate': 0.1, 'scale': 0.5, 'shear': 2.0, 'flipud': 0.0, 'fliplr': 0.5})
```

Feel free to adjust these parameters to suit your specific needs. If you encounter any issues, please ensure you are using the latest versions of the packages. For further assistance, providing a minimum reproducible example can be very helpful. You can find more details on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Happy training! 🚀

**Comment by shivam55sit** [comment]
This is great, thank you so much!

**Comment by glenn-jocher** [comment]
You're welcome! 😊

I'm glad you found the information helpful. If you have any further questions or run into any issues, feel free to reach out. Remember, if you encounter any problems, providing a minimum reproducible example can be very helpful for us to assist you better. You can find more details on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Also, make sure you're using the latest versions of the packages to avoid any known issues that might have been resolved in recent updates.

Happy fine-tuning and best of luck with your project! 🚀

**Comment by shivam55sit** [comment]
YOLOv8 models are better performing than many deep learning
state-of-the-art transfer learning models.

**Comment by glenn-jocher** [comment]
Thank you for your kind words! 😊

We're thrilled to hear that you find YOLOv8 models to be highly effective. If you have any specific questions or encounter any issues, please don't hesitate to reach out. Providing a minimum reproducible example can greatly assist us in diagnosing and resolving any problems you might face. You can find more details on creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, please ensure you're using the latest versions of our packages to benefit from the latest features and fixes. If you have any further questions or need assistance, feel free to ask. We're here to help!

Happy fine-tuning and best of luck with your project! 🚀

**Comment by Darshilgajera1** [comment]
Hi there, 

just need an guidance on how can I train a custom YOLOv8 model from which it can detect checkboxes from scanned image forms and give me text back. I am talking about here forms which has checkbox fields for example Gender has male, female, transgender checkbox.

**Comment by glenn-jocher** [comment]
@Darshilgajera1 to train a custom YOLOv8 model for detecting checkboxes on forms, you'll need to annotate a dataset of scanned forms with checkbox locations and train the model using this data. Ensure your dataset is diverse and well-labeled to improve detection accuracy.

**Comment by glenn-jocher** [comment]
@rafamartinssz you should use the `resume` parameter instead of loading `last.pt` as a new model. When you want to continue training from where you left off, use:

```python
from ultralytics import YOLO

# Resume training from checkpoint
model = YOLO()
results = model.train(resume='path/to/last.pt')
```

This properly restores the optimizer state, learning rate schedule, and epoch count, ensuring training continues seamlessly rather than restarting. The `resume` parameter is specifically designed for this use case and will maintain training continuity across sessions.
